src/Symfony/Component/Form/ChoiceList/Loader/FilterChoiceLoader.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\ChoiceList\Loader;
+
+use Symfony\Component\Form\ChoiceList\ArrayChoiceList;
+
+/**
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ */
+class FilterChoiceLoader implements ChoiceLoaderInterface
+{
+    private $loader;
+    private $filter;
+    private $choiceList;
+
+    public function __construct(ChoiceLoaderInterface $loader, callable $filter)
+    {
+        $this->loader = $loader;
+        $this->filter = $filter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoiceList($value = null)
+    {
+        if (null !== $this->choiceList) {
+            return $this->choiceList;
+        }
+
+        $choiceList = $this->loader->loadChoiceList($value);
+        $structured = $choiceList->getStructuredValues();
+        $choices = $choiceList->getChoices();
+        $visitor = function (array $list) use ($choices, &$visitor) {
+            foreach ($list as $k => $v) {
+                if (\is_array($v)) {
+                    if ($v = $visitor($v)) {
+                        $list[$k] = $v;
+                    } else {
+                        unset($list[$k]);
+                    }
+                    continue;
+                }
+
+                $choice = $choices[$v] ?? $v;
+                if (!\call_user_func($this->filter, $choice)) {
+                    unset($list[$k]);
+                    continue;
+                }
+
+                $list[$k] = $choice;
+            }
+
+            return $list;
+        };
+
+        return $this->choiceList = new ArrayChoiceList($visitor($structured), $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoicesForValues(array $values, $value = null)
+    {
+        // Optimize
+        if (empty($values)) {
+            return array();
+        }
+        if (null !== $this->choiceList) {
+            return $this->loadChoiceList($value)->getChoicesForValues($values);
+        }
+
+        return array_filter($this->loader->loadChoicesForValues($values, $value), $this->filter, ARRAY_FILTER_USE_BOTH);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadValuesForChoices(array $choices, $value = null)
+    {
+        // Optimize
+        if (empty($choices)) {
+            return array();
+        }
+        if (null !== $this->choiceList) {
+            return $this->loadChoiceList($value)->getValuesForChoices($choices);
+        }
+
+        return $this->loader->loadValuesForChoices(array_filter($choices, $this->filter, ARRAY_FILTER_USE_BOTH), $value);
```

Comment:
I'm quite sure this does not have the same arguments for the filter, as keys in this array have nothing to do with the original keys

Reply:
you're right. forgot about original keys :( then `$choice` only :) which should be only needed for filtering anyway :+1: